components/brave_wallet/browser/eth_data_parser.cc

```diff
@@ -80,66 +38,157 @@ bool GetTransactionInfoFromData(const std::string& data,
+  std::string calldata = data.substr(10);
+  if (selector == "0xa9059cbb") {
+    *tx_type = mojom::TransactionType::ERC20Transfer;
+    if (!brave_wallet::ABIDecode({"address", "uint256"}, calldata, tx_params,
+                                 tx_args)) {
       return false;
-    if (!GetUint256HexFromData(std::string(left_over_data), &value,
-                               &left_over_data))
+    }
+  } else if (selector == "0x095ea7b3") {
+    *tx_type = mojom::TransactionType::ERC20Approve;
+    if (!brave_wallet::ABIDecode({"address", "uint256"}, calldata, tx_params,
+                                 tx_args)) {
       return false;
-    // Very strictly must have correct data
-    if (left_over_data.length() > 0)
+    }
+  } else if (selector == "0x23b872dd") {
+    *tx_type = mojom::TransactionType::ERC721TransferFrom;
+    if (!brave_wallet::ABIDecode({"address", "address", "uint256"}, calldata,
+                                 tx_params, tx_args)) {
       return false;
-
-    if (tx_args) {
-      tx_args->push_back(address);
-      tx_args->push_back(value);
     }
-
-    if (tx_params) {
-      tx_params->push_back("address");
-      tx_params->push_back("uint256");
+  } else if (selector == "0x42842e0e") {
+    *tx_type = mojom::TransactionType::ERC721SafeTransferFrom;
+    if (!brave_wallet::ABIDecode({"address", "address", "uint256"}, calldata,
+                                 tx_params, tx_args)) {
+      return false;
     }
-  } else if (*tx_type == mojom::TransactionType::ERC721TransferFrom ||
-             *tx_type == mojom::TransactionType::ERC721SafeTransferFrom) {
-    std::string from, to, token_id;
-    std::string left_over_data = data.substr(10);
-    // Intentional copy of left_over_data
-    if (!GetAddressArgFromData(std::string(left_over_data), &from,
-                               &left_over_data))
+  } else if (selector == "0x3598d8ab") {
+    *tx_type = mojom::TransactionType::ETHSwap;
+
+    // Function:
+    // sellEthForTokenToUniswapV3(bytes encodedPath,
+    //                            uint256 minBuyAmount,
+    //                            address recipient)
+    //
+    // Ref:
+    // https://github.com/0xProject/protocol/blob/b46eeadc64485288add5940a210e1a7d0bcb5481/contracts/zero-ex/contracts/src/features/interfaces/IUniswapV3Feature.sol#L29-L41
+    if (!brave_wallet::ABIDecode({"bytes", "uint256", "address"}, calldata,
+                                 tx_params, tx_args)) {
       return false;
-    if (!GetAddressArgFromData(std::string(left_over_data), &to,
-                               &left_over_data))
+    }
+
+    std::vector<std::string> decoded_path;
+    if (!brave_wallet::UniswapEncodedPathDecode(tx_args->at(0), decoded_path))
       return false;
-    if (!GetUint256HexFromData(std::string(left_over_data), &token_id,
-                               &left_over_data))
+    std::string fill_path = "0x";
+    for (const auto& path : decoded_path) {
+      fill_path += path.substr(2);
```

Comment:
please use `base::StrAppend` which avoids a temporary string allocation and copy.
Same rule applies to https://github.com/brave/brave-core/pull/12797/files#diff-832b548af9f742c2cd91f0f70f8e85102e93c3ca4ded730672f894b1d139d580R157
and https://github.com/brave/brave-core/pull/12797/files#diff-d41475c64c412cc9e28690a6895d70b97fa1ef66ce4dbdb5ce865d89ccc760daR130

Reply:
Fixed here: 04b86cff3fdf2177b5f3aa56b05c4f21c7ce2a25, along with few other string optimizations.